src/audio/module_adapter/module/iadk_modules.c

```diff
@@ -125,8 +160,15 @@ static int iadk_modules_prepare(struct processing_module *mod)
 	comp_info(dev, "iadk_modules_prepare()");
 
 	/* Call module specific prepare function if exists. */
-	ret = iadk_wrapper_prepare(mod->priv.module_adapter);
-	return 0;
+	if (mod->is_loadable) {
+		struct module_interface *mod_in =
+					(struct module_interface *)mod->priv.module_adapter;
+
+		ret = mod_in->prepare(mod);
```

Comment:
Don't we have module API wrappers here that check for NULL ptrs before the dereference ?

Reply:
```ret = module_adapter_prepare(mod);``` ?

Reply:
In SOF modules we don't you module adapter but we call modules directly without component api wrapper.

Reply:
I'm not sure about check for NULL ptrs. There are some macros calling priv data but not suitable for SOF modules.

Reply:
Not understanding why we dont need to call the module API and use the op ptr directly here ?
The call does a lot more than cal the op. In fact, it does need to check the op ptr before use if prepare() op is optional.

```c
int module_prepare(struct processing_module *mod)
{
	int ret;
	struct module_data *md = &mod->priv;
	struct comp_dev *dev = mod->dev;

	comp_dbg(dev, "module_prepare() start");

	if (mod->priv.state == MODULE_IDLE)
		return 0;
	if (mod->priv.state < MODULE_INITIALIZED)
		return -EPERM;

	ret = md->ops->prepare(mod);
	if (ret) {
		comp_err(dev, "module_prepare() error %d: module specific prepare failed, comp_id %d",
			 ret, dev_comp_id(dev));
		return ret;
	}

	/* After prepare is done we no longer need runtime configuration
	 * as it has been applied during the procedure - it is safe to
	 * free it.
	 */
	if (md->cfg.data)
		rfree(md->cfg.data);

	md->cfg.avail = false;
	md->cfg.data = NULL;

	md->state = MODULE_IDLE;
	comp_dbg(dev, "module_prepare() done");

	return ret;
}
```

Reply:
"Not understanding why we dont need to call the module API and use the op ptr directly here ?" Because that op is delivered by module itself. Mechanism is the same as
`static struct module_interface up_down_mixer_interface = {
	.init  = up_down_mixer_init,
	.prepare = up_down_mixer_prepare,
	.process = up_down_mixer_process,
	.reset = up_down_mixer_reset,
	.free = up_down_mixer_free
};`

Reply:
Sorry, I'm confused. This change bypasses the state management in the module API by calling the op directly.
The module adpater should be the base class here and the iadk the derived class. i.e. the extra features the iadk provides are built on top of the module adaptor and NOT alongside the module adaptor.

Reply:
@pjdobrowolski what you are missing is the call to module_init() instead of mod_in->init() Once you do this you can call all the other wrappers such as module_prepare() as well, right?

Reply:
@pjdobrowolski any update here ? iadk_modules must call the lower base class module API. This ensure we have conssitency across all state machines and de-duplication..

Reply:
Not sure @pjdobrowolski you are talking about different things, but I tried to recap, and it would seem this is still a valid concern. When module adapter is used with built-in modules, all calls go via common wrapppers (e.g. module_prepare()). I don't see why we wouldn't use these common wrappers also when the module is loaded.

I mean, e.g. for upmix, it implements prepare() for sure, but when it is built in, we still call upmix prepare via module_prepare(), right?